### PR TITLE
Advice injector refactoring

### PR DESCRIPTION
**Pull request by bobbinth**
This PR refactors advice injector names, handling logic, and introduces a couple new advice injectors. Specifically:

- [x] Split advice injector handlers int the processor into multiple files.
- [x] Cleaned up `ExecutionError` enum a bit (i.e., broke out `Ext2InttError` into a separate sub-enum).
- [x] Renamed advice instructions to have the form `adv.push_xxx` for instructions which push data onto the advice stack and `adv.insert_xxx` for instructions which insert data into the advice map.
- [x] Added 2 new advice injectors and removed one injector which does not need to be exposed to the user (see detailed list below).
- [x] Update user docs.

Instruction set updates (some of these are breaking changes):
- Renamed instructions:
  - `adv.u64div` -> `adv.push_u64div`.
  - `adv.ext2intt` -> `adv.push_ext2intt`.
  - `adv.smtget` -> `adv.push_smtget`.
  - `adv.keyval` -> `adv.push_mapval`.
  - `adv.mem` -> `adv.insert_mem`.
- Removed instructions:
  - `adv.ext2inv`.
- Added instructions:
  - `adv.push_mtnode` for non-deterministically getting a value of a Merkle tree node.
  - `adv.insert_hdword` for inserting top two words from the stack into the advice map.

### CODE REVIEW

File: assembly/src/ast/nodes/advice.rs [hunk 1]
```diff
@@ -9,52 +10,92 @@ use core::fmt;
...
+// SERIALIZATION / DESERIALIZATION
+// ================================================================================================
+
+const PUSH_U64DIV: u8 = 0;
+const PUSH_EXT2INTT: u8 = 1;
+const PUSH_SMTGET: u8 = 2;
+const PUSH_MAPVAL: u8 = 3;
+const PUSH_MTNODE: u8 = 4;
+const INSERT_MEM: u8 = 5;
+const INSERT_HDWORD: u8 = 6;
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply assembly/src/ast/nodes/advice.rs:66]: The main reason is that here we have very few variants (for now) and this approach seems more readable (I did implement the same approach as with `OpCode` at first - but it was more code and actually hurt readability). I think once the number of variants here grows, we'll need to transition to something which is more like `OpCode`. Though, to be honest, I don't love that approach either and wondering if we can come up with something better.

File: core/src/operations/decorators/advice.rs [hunk 2]
```diff
@@ -1,65 +1,101 @@
...
+    // --------------------------------------------------------------------------------------------
     /// Creates a new Merkle tree in the advice provider by combining Merkle trees with the
-    /// specified roots. The root of the new tree is defined as `Hash(left_root, right_root)`.
+    /// specified roots. The root of the new tree is defined as `Hash(LEFT_ROOT, RIGHT_ROOT)`.
     ///
     /// The operand stack is expected to be arranged as follows:
-    /// - root of the right tree, 4 elements.
-    /// - root of the left tree, 4 elements.
+    ///
+    /// [RIGHT_ROOT, LEFT_ROOT, ...]
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/operations/decorators/advice.rs:24]: Agreed! Updated.

File: assembly/src/ast/nodes/advice.rs [hunk 3]
```diff
@@ -9,52 +10,92 @@ use core::fmt;
...
+            AdviceInjector::PushExt2intt => Self::Ext2Intt,
+            AdviceInjector::PushSmtGet => Self::SmtGet,
+            AdviceInjector::PushMapVal => Self::MapValueToStack,
+            AdviceInjector::PushMtNode => Self::MerkleNodeToStack,
+            AdviceInjector::InsertMem => Self::MemToMap,
+            AdviceInjector::InsertHdword { domain } => Self::HdwordToMap {
+                domain: Felt::from(*domain),
+            },
+        }
+    }
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply assembly/src/ast/nodes/advice.rs:37]: Renamed to `AdviceInjectorNode`.

File: processor/src/errors.rs [hunk 4]
```diff
@@ -17,24 +17,22 @@ use std::error::Error;
...
     InvalidFriLayerFolding(QuadFelt, QuadFelt),
+    InvalidMemoryRange { start_addr: u64, end_addr: u64 },
     InvalidStackDepthOnReturn(usize),
-    NttDomainSizeTooSmall(u64),
-    NttDomainSizeNotPowerOf2(u64),
-    InterpolationResultSizeTooBig(usize, usize),
+    InvalidTreeDepth { depth: Felt },
+    InvalidTreeNodeIndex { depth: Felt, value: Felt },
+    MemoryAddressOutOfBounds(u64),
+    MerkleStoreMergeFailed(MerkleError),
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/errors.rs:33]: Yep - that's an issue. To solve this we should also update slightly how MMR works. I filed https://github.com/0xPolygonMiden/miden-vm/issues/943 for this.

**Comment by bobbinth** [comment]
Addressed comments and also added docs to mdBook documentation for all advice injectors.

**Review by jjcnn — APPROVED** [review]
LGTM

File: core/src/operations/decorators/advice.rs [hunk 5]
```diff
@@ -1,86 +1,189 @@
...
 
 impl fmt::Display for AdviceInjector {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
-            Self::MerkleNode => write!(f, "merkle_node"),
-            Self::MerkleMerge => write!(f, "merkle_merge"),
-            Self::DivResultU64 => write!(f, "div_result_u64"),
-            Self::MapValue => write!(f, "map_value"),
-            Self::Memory => write!(f, "mem"),
+            Self::MerkleNodeMerge => write!(f, "merkle_tree_merge"),
```
Comment (frisitano) [inline core/src/operations/decorators/advice.rs:178]: Should we also update the display to be `merkle_node_merge`?

**Review by frisitano — APPROVED** [review]
Looks good! Left a non blocking comment inline.

File: core/src/operations/decorators/advice.rs (hunk 5 above)
Comment (bobbinth) [reply core/src/operations/decorators/advice.rs]: Good catch! Updated.

File: processor/src/decorators/mod.rs [hunk 6]
```diff
@@ -57,321 +35,46 @@ where
...
         match injector {
-            AdviceInjector::MerkleNode => self.inject_merkle_node(),
-            AdviceInjector::MerkleMerge => self.inject_merkle_merge(),
-            AdviceInjector::DivResultU64 => self.inject_div_result_u64(),
-            AdviceInjector::MapValue => self.inject_map_value(),
-            AdviceInjector::Memory => self.inject_mem_values(),
-            AdviceInjector::Ext2Inv => self.inject_ext2_inv_result(),
-            AdviceInjector::Ext2INTT => self.inject_ext2_intt_result(),
-            AdviceInjector::SmtGet => self.inject_smtget(),
+            AdviceInjector::MerkleNodeMerge => self.merge_merkle_trees(),
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decorators/mod.rs]: I'd prefer to keep it verb-based - so, changed to `merge_merkle_nodes()`.

**Pull request merged**